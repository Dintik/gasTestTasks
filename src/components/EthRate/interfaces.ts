export interface IRateEth {
    ethRate: null | string,
    isLoadingEth: boolean,
};