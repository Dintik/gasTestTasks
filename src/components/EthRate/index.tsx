import React, { useEffect, useCallback } from "react";
import { useAppSelector, useAppDispatch } from "../../store/hooks";
import { setEthRate, setLoadingEth } from "../../store/prices/reducer";
import styles from "./styles.module.scss";
import { getEthRate } from "../../api/getEthRate";
import { IRateEth } from "./interfaces";

const EthRate: React.FC = () => {
    const price: IRateEth = useAppSelector((state) => state.pricesState);
    const dispatch = useAppDispatch();

    const refreshData = useCallback(async function () {
        dispatch(setLoadingEth());
        const curentEthRate = await getEthRate();
        dispatch(setEthRate(curentEthRate));
    }, []);

    useEffect(() => {
        refreshData();
    }, []);

    return (
        <section className={styles.ethRate}>
            <button className={styles.ethRate__btn} type="button" onClick={refreshData}>
                Refresh Eth Rate
            </button>
            <div className={styles.ethRate__text__wrap}>
                <p className={styles.ethRate__text}>Exchange rate: {price.isLoadingEth ? "Loading..." : `${price.ethRate}$`}</p>
            </div>
        </section>
    );
};

export default EthRate;
