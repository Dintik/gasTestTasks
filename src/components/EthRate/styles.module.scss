.ethRate {
    text-align: center;
    display: flex;
    align-items: center;
    justify-content: center;
    font-size: calc(10px + 2vmin);

    &__text {
        font-size: calc(10px + 2vmin);
        margin: 20px;
        min-width: 370px;
        text-align: left;
    }

    &__btn {
        font-size: calc(10px + 2vmin);
        margin: 10px;
        padding: 20px 30px;
        min-width: 345px;
    }
}
