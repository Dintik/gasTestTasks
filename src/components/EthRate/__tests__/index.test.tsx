import thunk from "redux-thunk";
import configureMockStore from "redux-mock-store";
import { Provider } from "react-redux";
import { render } from "@testing-library/react";
import EthRate from "../index";
const mockStore = configureMockStore([thunk]);

describe("Gas Price", () => {
    it("should render loading", () => {
        const store = mockStore({
            pricesState: {
                gasPrice: null,
                ethRate: null,
                isLoadingGas: true,
                isLoadingEth: true
            }
        });

        const { asFragment } = render(
            <Provider store={store}>
                <EthRate />
            </Provider>
        );

        expect(asFragment()).toMatchSnapshot();
    });

    it("should render correctly without data", () => {
        const store = mockStore({
            pricesState: {
                ethRate: "Sorry, there was an error",
                gasPrice: "Sorry, there was an error",
                isLoadingGas: false,
                isLoadingEth: false
            }
        });

        const { asFragment } = render(
            <Provider store={store}>
                <EthRate />
            </Provider>
        );

        expect(asFragment()).toMatchSnapshot();
    });

    it("should render data", () => {
        const store = mockStore({
            pricesState: {
                ethRate: "2697.51",
                gasPrice: "154.788",
                isLoadingGas: false,
                isLoadingEth: false
            }
        });

        const { asFragment } = render(
            <Provider store={store}>
                <EthRate />
            </Provider>
        );

        expect(asFragment()).toMatchSnapshot();
    });
});

export {};
