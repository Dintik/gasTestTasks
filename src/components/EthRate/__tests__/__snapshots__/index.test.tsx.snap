// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gas Price should render correctly without data 1`] = `
<DocumentFragment>
  <section
    class="ethRate"
  >
    <button
      class="ethRate__btn"
      type="button"
    >
      Refresh Eth Rate
    </button>
    <div
      class="ethRate__text__wrap"
    >
      <p
        class="ethRate__text"
      >
        Exchange rate: Sorry, there was an error$
      </p>
    </div>
  </section>
</DocumentFragment>
`;

exports[`Gas Price should render data 1`] = `
<DocumentFragment>
  <section
    class="ethRate"
  >
    <button
      class="ethRate__btn"
      type="button"
    >
      Refresh Eth Rate
    </button>
    <div
      class="ethRate__text__wrap"
    >
      <p
        class="ethRate__text"
      >
        Exchange rate: 2697.51$
      </p>
    </div>
  </section>
</DocumentFragment>
`;

exports[`Gas Price should render loading 1`] = `
<DocumentFragment>
  <section
    class="ethRate"
  >
    <button
      class="ethRate__btn"
      type="button"
    >
      Refresh Eth Rate
    </button>
    <div
      class="ethRate__text__wrap"
    >
      <p
        class="ethRate__text"
      >
        Exchange rate: Loading...
      </p>
    </div>
  </section>
</DocumentFragment>
`;
